README.md
foundationdb Cookbook
=====================
This cookbook installs and configure FoundationDB.

Requirements
------------
<none>

Attributes
----------
<table>
  <tr>
    <th>Key</th>
    <th>Type</th>
    <th>Description</th>
    <th>Default</th>
  </tr>
  <tr>
    <td><tt>['foundationdb']['role']</tt></td>
    <td>String</td>
    <td>Install server or client. Possible values: client, server</td>
    <td><tt>'server'</tt></td>
  </tr>
  <tr>
    <td><tt>['foundationdb']['install_type']</tt></td>
    <td>String</td>
    <td>Full installation (configurations as well) or upgrade.
        Possible values: full, upgrade</td>
    <td><tt>'full'</tt></td>
  </tr>
  <tr>
    <td><tt>['foundationdb']['version']</tt></td>
    <td>String</td>
    <td>FoundationDB version to install.  There will soon be support for the
        value 'latest', but not today ):</td>
    <td><tt>'2.0.10'</tt></td>
  </tr>
  <tr>
    <td><tt>['foundationdb']['data_dir']</tt></td>
    <td>String or nil</td>
    <td>Path to data directory for storage.  A value of <tt>nil</tt> will
        result in the FoundationDB default location.</td>
    <td><tt>nil</tt></td>
  </tr>
  <tr>
    <td><tt>['foundationdb']['log_dir']</tt></td>
    <td>String or nil</td>
    <td>Path to directory for logs.  A value of <tt>nil</tt> will
        result in the FoundationDB default location.</td>
    <td><tt>nil</tt></td>
  </tr>
  <tr>
    <td><tt>['foundationdb']['datacenter_id']</tt></td>
    <td>Hexadecimal String or nil</td>
    <td>16-character hexadecimal string denoting the datacenter ID.  A value
        of nil will leave this parameter unset.</td>
    <td><tt>nil</tt></td>
  </tr>
  <tr>
    <td><tt>['foundationdb']['processes']</tt></td>
    <td>List of dictionaries</td>
    <td>Only 'port' is required. Format:<br />
      <pre>
[
  {
    'port' => 4500,
    'id' => 'string',
    'file' => '/etc/foundationdb/file.cluster',
    'memory' => 24,
    'storage_memory' => 16,
    'count' => 1
  },
  .
  .
  {
    ..
  }
]
      </pre>
    </td>
    <td><pre>
[
  {
    'port' => 4500
  }
]
    </tt></pre>
  </tr>
  <tr>
    <td><tt>['foundationdb']['base_url']</tt></td>
    <td>String</td>
    <td>URL base path from which the FoundationDB packages can be downloaded.
        This makes it easier to specify private locations if the files are
        being managed internally.</td>
    <td><tt>'https://foundationdb.com/downloads/I_accept_the_FoundationDB_Community_License_Agreement'</tt></td>
  </tr>
  <tr>
    <td><tt>['foundationdb']['make_public']</tt></td>
    <td>Boolean</td>
    <td>Run make_public after installation.  Possible values: true, false</td>
    <td><tt>true</tt></td>
  </tr>
  <tr>
    <td><tt>['foundationdb']['cleanup']</tt></td>
    <td>Boolean</td>
    <td>When true, removes any temporary files that were needed to set up the
        new version (.rpm files, .deb files, etc.).</td>
    <td><tt>true</tt></td>
  </tr>
  <tr>
    <td><tt>['foundationdb']['temp_dir']</tt></td>
    <td>Path String</td>
    <td>Location for temporary files needed to install FoundationDB.</td>
    <td><tt>/tmp</tt></td>
  </tr>
</table>
There are other parameters that are being moved internal so that users
will not have to worry about them.

Usage
-----
Just include `foundationdb` in your node's `run_list`:
```json
{
  "name": "my_node",
  "run_list": [
    "recipe[foundationdb]"
  ]
}
```

Contributing
------------
1. Fork the repository on Github
2. Create a named feature branch (like `add_component_x`)
3. Write your change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request using Github

License and Authors
-------------------
Authors:
  Kevin (penniesfromkevin)


CHANGELOG.md
foundationdb CHANGELOG
======================

This file is used to list changes made in each version of the foundationdb cookbook.

0.5.0
-----
- Kevin - Hard-coded deb packages due to FoundationDB binaries no longer being generally available.

0.4.2
-----
- Kevin - Trying force-confold for upgrades.

0.4.1
-----
- Kevin - Fixed server_source_url link.

0.4.0
-----
- Kevin - 1) Added direct URL attributes for packages.
          2) Eliminated some dubious attributes.

0.3.1
-----
- Kevin - Cookbook reference fixed.

0.2.7
-----
- Kevin - Initial public release of foundationdb cookbook

0.1.0
-----
- Kevin - Initial checkin of foundationdb cookbook

- - -
Check the [Markdown Syntax Guide](http://daringfireball.net/projects/markdown/syntax) for help with Markdown.

The [Github Flavored Markdown page](http://github.github.com/github-flavored-markdown/) describes the differences between markdown on github and standard markdown.


recipes/client.rb
#
# Cookbook Name:: foundationdb
# Recipe:: client
#
# Copyright (c) 2014-2015, Kevin (penniesfromkevin)
#
# All rights reserved - Do Not Redistribute
#
Chef::Log.info 'Installing FoundationDB client'
if node['foundationdb']['clients_url']
  clients_file = node['foundationdb']['clients_url'].split("/").last
  clients_temp_file = "/tmp/#{clients_file}"

  remote_file clients_temp_file do
    source node['foundationdb']['clients_url']
    mode 00755
  end
else
  clients_file = "foundationdb-clients_#{node['foundationdb']['version']}-1_amd64.deb"
  clients_temp_file = "/tmp/#{clients_file}"

  cookbook_file clients_file do
    path clients_temp_file
    mode 00755
    action :create
  end
end

dpkg_package 'foundationdb-clients' do
  source clients_temp_file
  action [:install]
end

directory '/etc/foundationdb' do
  mode 00777
  action :create
end

# Delete temp files
file clients_temp_file do
  action :delete
end


attributes/defaults.rb
#
# Copyright (c) 2014-2015, Kevin (penniesfromkevin)
#
# Installations can be full or upgrade only (does not touch configurations)
#default['foundationdb']['install_type'] = 'upgrade'
default['foundationdb']['install_type'] = 'full'

# ssd processes
default['foundationdb']['processes'] = [
  {
    'port' => 4500
  }
]

# Optional datacenter ID
default['foundationdb']['datacenter_id'] = nil

# These will change the default locations
default['foundationdb']['data_dir'] = nil
default['foundationdb']['log_dir'] = nil

# Explicit package URLs, if available
default['foundationdb']['clients_url'] = nil
default['foundationdb']['server_url'] = nil

default['foundationdb']['make_public'] = true
# When make_public accepts different options, they can be specified here
default['foundationdb']['make_public_options'] = ''

# Manage the versions if explicit package URLs have not been provided.
default['foundationdb']['version'] = '2.0.10'


recipes/default.rb
#
# Cookbook Name:: foundationdb
# Recipe:: default
#
# Installs FoundationDB server.
#
# Copyright (c) 2014-2015, Kevin (penniesfromkevin)
#
# All rights reserved - Do Not Redistribute
#
# Client must be installed before server(!)
include_recipe 'foundationdb::client'

Chef::Log.info 'Installing FoundationDB server'
if node['foundationdb']['server_url']
  server_file = node['foundationdb']['server_url'].split("/").last
  server_temp_file = "/tmp/#{server_file}"

  remote_file server_temp_file do
    source node['foundationdb']['server_url']
    mode 00755
  end
else
  server_file = "foundationdb-server_#{node['foundationdb']['version']}-1_amd64.deb"
  server_temp_file = "/tmp/#{server_file}"

  cookbook_file server_file do
    path server_temp_file
    mode 00755
    action :create
  end
end

# See if FoundationDB has been previously installed (fdb.cluster exists)
initial_file = '/etc/foundationdb/fdb.cluster'
if File.exists?(initial_file)
  initial_install = false
else
  initial_install = true
end

dpkg_package 'foundationdb-server' do
  options '--force-confold'
  source server_temp_file
  action [:install]
end

if node['foundationdb']['install_type'] == 'full'
  foundationdb_conf = '/etc/foundationdb/foundationdb.conf'

  if node['foundationdb']['data_dir']
    Chef::Log.info 'Configuring FoundationDB data_dir'
    directory node['foundationdb']['data_dir'] do
      owner 'foundationdb'
      group 'foundationdb'
      mode 00777
      recursive true
      action :create
    end

    ruby_block 'Edit foundationdb.conf data_dir' do
      block do
        rc = Chef::Util::FileEdit.new(foundationdb_conf)
        rc.search_file_replace_line("^datadir = /var/lib/foundationdb/data/.*", "datadir = #{node['foundationdb']['data_dir']}/$ID")
        rc.write_file
      end
    end
  end

  if node['foundationdb']['log_dir']
    Chef::Log.info 'Configuring FoundationDB log_dir'
    directory node['foundationdb']['log_dir'] do
      owner 'foundationdb'
      group 'foundationdb'
      mode 00777
      recursive true
      action :create
    end

    ruby_block 'Edit foundationdb.conf log_dir' do
      block do
        rc = Chef::Util::FileEdit.new(foundationdb_conf)
        rc.search_file_replace_line('^logdir = /var/log/foundationdb', "logdir = #{node['foundationdb']['log_dir']}")
        rc.write_file
      end
    end
  end

  if node['foundationdb']['datacenter_id']
    ruby_block 'Edit foundationdb.conf datacenter_id' do
      block do
        rc = Chef::Util::FileEdit.new(foundationdb_conf)
        rc.search_file_replace_line('# datacenter_id =.*', 'datacenter_id = ')
        rc.write_file
        rc.search_file_replace_line('datacenter_id =.*', "datacenter_id = #{node['foundationdb']['datacenter_id']}")
        rc.write_file
      end
    end
  end

  if node['foundationdb']['processes']
    # Comment out the initial process; user may not want 4500
    ruby_block 'Edit foundationdb.conf fdbserver.port' do
      block do
        rc = Chef::Util::FileEdit.new(foundationdb_conf)
        rc.search_file_replace_line("^.fdbserver\.4500.", "##[fdbserver.4500]")
        rc.write_file
      end
    end
  end

  # Create new processes
  node['foundationdb']['processes'].each do |process|
    fdb_port = process['port']
    if process.has_key?(:count)
      fdb_count = process['count']
    else
      fdb_count = 1
    end
    process_options = ''
    if process.has_key?(:id)
      process_id = process['id']
    else
      process_id = 'local'
    end
    if process.has_key?(:string)
      process_string = process['string']
    else
      process_string = 'fdbstring'
    end
    if process.has_key?(:file)
      # Processes with files are usually memory processes
      file process['file'] do
        content "#{process_id}:#{process_string}@#{node['ipaddress']}:#{fdb_port}"
        owner 'foundationdb'
        group 'foundationdb'
        mode 00666
        action :create_if_missing
      end
      process_options += "\ncluster_file = #{process['file']}"
    elsif initial_install and not process.has_key?(:class)
      # Processes without files are SSD processes
      file initial_file do
        content "#{process_id}:#{process_string}@#{node['ipaddress']}:#{fdb_port}"
        owner 'foundationdb'
        group 'foundationdb'
        mode 00666
        action :create
      end
      # Only do this once
      initial_install = false
    end
    if process.has_key?(:memory)
      process_options += "\nmemory = #{process['memory']}GiB"
    end
    if process.has_key?(:storage_memory)
      process_options += "\nstorage_memory = #{process['storage_memory']}GiB"
    end
    if process.has_key?(:class)
      process_options += "\nclass = #{process['class']}"
    end
    ruby_block "Edit foundationdb.conf for process settings #{fdb_port}" do
      block do
        (1..fdb_count).each do|i|
          rc = Chef::Util::FileEdit.new(foundationdb_conf)
          rc.insert_line_if_no_match("^.fdbserver\.#{fdb_port}.", "\n# Process port #{fdb_port}\n[fdbserver.#{fdb_port}]#{process_options}")
          rc.write_file
          fdb_port += 1
        end
      end
    end
  end

  if node['foundationdb']['make_public']
    Chef::Log.info 'Making FoundationDB node public'
    execute 'make public' do
      command "/usr/lib/foundationdb/make_public.py #{node['foundationdb']['make_public_options']}"
      action :run
      # temporary until idempotency is implemented
      ignore_failure true
    end
  end
end

# Delete temp files
file server_temp_file do
  action :delete
end
